// graphql-query-to-json/dist/test/full_functionality.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("../index");
var json_to_graphql_query_1 = require("json-to-graphql-query");
describe("Queries", function () {
    it("Single property", function () {
        var query = "\n            query {\n                viewer {\n                    theOnlyPropertyIWant\n                }\n                other {\n                    anotherOne\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    theOnlyPropertyIWant: true,
                },
                other: {
                    anotherOne: true,
                },
            },
        });
    });
    it("Two properties", function () {
        var query = "\n            query {\n                viewer {\n                    propertyOne\n                    propertyTwo\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    propertyOne: true,
                    propertyTwo: true,
                },
            },
        });
    });
    it("Two properties separated by commas", function () {
        var query = "\n            query {\n                viewer {\n                    propertyOne,\n                    propertyTwo\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    propertyOne: true,
                    propertyTwo: true,
                },
            },
        });
    });
    it("Nested simple query using commas", function () {
        var query = "\n            query {\n                viewer {\n                    propertyOne,\n                    propertyTwo\n                }\n                nested {\n                    evenDeeper {\n                        nestedOne,\n                        nestedTwo,\n                        nestedThree\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    propertyOne: true,
                    propertyTwo: true,
                },
                nested: {
                    evenDeeper: {
                        nestedOne: true,
                        nestedTwo: true,
                        nestedThree: true,
                    },
                },
            },
        });
    });
    it("Nested simple query using commas with custom query name", function () {
        var query = "\n            query GetThatStuff {\n                viewer {\n                    propertyOne,\n                    propertyTwo\n                }\n                nested {\n                    evenDeeper {\n                        nestedOne,\n                        nestedTwo,\n                        nestedThree\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    propertyOne: true,
                    propertyTwo: true,
                },
                nested: {
                    evenDeeper: {
                        nestedOne: true,
                        nestedTwo: true,
                        nestedThree: true,
                    },
                },
            },
        });
    });
    it("Simple query using variables", function () {
        var query = "\n        query GetThisStuff($name: String, $lastName: String) {\n            viewer {\n                personal(criteria: {\n                    name: $name,\n                    lastName: $lastName\n                }) {\n                    name\n                    address\n                }\n            }\n        }\n        ";
        var result = (0, index_1.graphQlQueryToJson)(query, {
            variables: {
                name: "PETER",
                lastName: "SCHMIDT",
            },
        });
        expect(result).toEqual({
            query: {
                viewer: {
                    personal: {
                        __args: {
                            criteria: {
                                name: "PETER",
                                lastName: "SCHMIDT",
                            },
                        },
                        name: true,
                        address: true,
                    },
                },
            },
        });
    });
    it("Scalar field with arguments", function () {
        var query = "\n        query fooBarAlias($name: String!, $lastName: String!) {\n            fooBar(name: $name, lastName: $lastName)\n        }";
        var result = (0, index_1.graphQlQueryToJson)(query, {
            variables: {
                name: "PETER",
                lastName: "SCHMIDT",
            },
        });
        expect(result).toEqual({
            query: {
                fooBar: {
                    __args: {
                        name: "PETER",
                        lastName: "SCHMIDT",
                    },
                },
            },
        });
    });
});
describe("Mutations", function () {
    it("Emtpy properties", function () {
        var mutation = "\n            mutation {\n                getPersonalStuff (info: {\n                    emptyString: \"\",\n                    emptyObject: {},\n                    emptyArray: [],\n                }) {\n                    personal\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(mutation)).toEqual({
            mutation: {
                getPersonalStuff: {
                    personal: true,
                    __args: {
                        info: {
                            emptyString: "",
                            emptyObject: {},
                            emptyArray: [],
                        },
                    },
                },
            },
        });
    });
    it("Simple mutation using string argument with sibling queries", function () {
        var mutation = "\n        mutation {\n            getPersonalStuff(name: \"PETER\") {\n                personal {\n                    name\n                    address\n                }\n                other {\n                    parents\n                }\n            }\n        }\n        ";
        expect((0, index_1.graphQlQueryToJson)(mutation)).toEqual({
            mutation: {
                getPersonalStuff: {
                    __args: {
                        name: "PETER",
                    },
                    personal: {
                        name: true,
                        address: true,
                    },
                    other: {
                        parents: true,
                    },
                },
            },
        });
    });
    it("Simple mutation using multiple string arguments with sibling queries", function () {
        var mutation = "\n        mutation {\n            getPersonalStuff(name: \"AMADEUS\", lastName: \"MOZART\") {\n                personal {\n                    name\n                    address\n                }\n                other {\n                    parents\n                }\n            }\n        }\n        ";
        expect((0, index_1.graphQlQueryToJson)(mutation)).toEqual({
            mutation: {
                getPersonalStuff: {
                    __args: {
                        name: "AMADEUS",
                        lastName: "MOZART",
                    },
                    personal: {
                        name: true,
                        address: true,
                    },
                    other: {
                        parents: true,
                    },
                },
            },
        });
    });
    it("Arguments wrapped in object", function () {
        var mutation = "\n        mutation {\n            getPersonalStuff(input: {\n                name: \"HANNES\",\n                lastName: \"RUDOLF\"\n            }) {\n                personal {\n                    name\n                    address\n                }\n                other {\n                    parents\n                }\n            }\n        }\n        ";
        expect((0, index_1.graphQlQueryToJson)(mutation)).toEqual({
            mutation: {
                getPersonalStuff: {
                    __args: {
                        input: {
                            name: "HANNES",
                            lastName: "RUDOLF",
                        },
                    },
                    personal: {
                        name: true,
                        address: true,
                    },
                    other: {
                        parents: true,
                    },
                },
            },
        });
    });
    it("Arguments wrapped in object using an int value", function () {
        var mutation = "\n        mutation {\n            getPersonalStuff(input: {\n                count: 1000,\n            }) {\n                personal {\n                    name\n                    address\n                }\n                other {\n                    parents\n                }\n            }\n        }\n        ";
        expect((0, index_1.graphQlQueryToJson)(mutation)).toEqual({
            mutation: {
                getPersonalStuff: {
                    __args: {
                        input: {
                            count: 1000,
                        },
                    },
                    personal: {
                        name: true,
                        address: true,
                    },
                    other: {
                        parents: true,
                    },
                },
            },
        });
    });
    it("Arguments wrapped in nested object", function () {
        var mutation = "\n        mutation {\n            getPersonalStuff(input: {\n                name: \"HANNES\",\n                lastName: \"RUDOLF\",\n                city: {\n                    name: \"Hollywood\",\n                    country: \"California, USA\"\n                }\n            }) {\n                personal {\n                    name\n                    address\n                }\n                other {\n                    parents\n                }\n            }\n        }\n        ";
        expect((0, index_1.graphQlQueryToJson)(mutation)).toEqual({
            mutation: {
                getPersonalStuff: {
                    __args: {
                        input: {
                            name: "HANNES",
                            lastName: "RUDOLF",
                            city: {
                                name: "Hollywood",
                                country: "California, USA",
                            },
                        },
                    },
                    personal: {
                        name: true,
                        address: true,
                    },
                    other: {
                        parents: true,
                    },
                },
            },
        });
    });
    it("Argument is a list", function () {
        var mutation = "\n        mutation {\n            getPersonalStuff(input: [\n                \"ONE\",\n                \"TWO\"\n            ]) {\n                personal {\n                    name\n                    address\n                }\n                other {\n                    parents\n                }\n            }\n        }\n        ";
        expect((0, index_1.graphQlQueryToJson)(mutation)).toEqual({
            mutation: {
                getPersonalStuff: {
                    __args: {
                        input: ["ONE", "TWO"],
                    },
                    personal: {
                        name: true,
                        address: true,
                    },
                    other: {
                        parents: true,
                    },
                },
            },
        });
    });
    it("Argument is a int", function () {
        var mutation = "\n        mutation {\n            getUsers(count: 1000) {\n                personal {\n                    count\n                }\n            }\n        }";
        expect((0, index_1.graphQlQueryToJson)(mutation)).toEqual({
            mutation: {
                getUsers: {
                    __args: {
                        count: 1000,
                    },
                    personal: {
                        count: true,
                    },
                },
            },
        });
    });
});
describe("Subscriptions", function () {
    it("Simple subscription", function () {
        var subscription = "\n            subscription {\n                messageAdded {\n                    id\n                    content\n                    user {\n                        name\n                        email\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(subscription)).toEqual({
            subscription: {
                messageAdded: {
                    id: true,
                    content: true,
                    user: {
                        name: true,
                        email: true,
                    },
                },
            },
        });
    });
    it("Subscription with arguments", function () {
        var subscription = "\n            subscription MessageSubscription($userId: ID!) {\n                messageAdded(userId: $userId) {\n                    id\n                    content\n                    timestamp\n                }\n            }\n        ";
        var result = (0, index_1.graphQlQueryToJson)(subscription, {
            variables: {
                userId: "123",
            },
        });
        expect(result).toEqual({
            subscription: {
                messageAdded: {
                    __args: {
                        userId: "123",
                    },
                    id: true,
                    content: true,
                    timestamp: true,
                },
            },
        });
    });
    it("Subscription with alias", function () {
        var subscription = "\n            subscription {\n                latestMessage: messageAdded {\n                    id\n                    content\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(subscription)).toEqual({
            subscription: {
                latestMessage: {
                    __aliasFor: "messageAdded",
                    id: true,
                    content: true,
                },
            },
        });
    });
    it("Subscription with enum arguments", function () {
        var subscription = "\n            subscription {\n                messageAdded(channel: PUBLIC, priority: HIGH) {\n                    id\n                    content\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(subscription)).toEqual({
            subscription: {
                messageAdded: {
                    __args: {
                        channel: new json_to_graphql_query_1.EnumType("PUBLIC"),
                        priority: new json_to_graphql_query_1.EnumType("HIGH"),
                    },
                    id: true,
                    content: true,
                },
            },
        });
    });
    it("Subscription with scalar field with arguments", function () {
        var subscription = "\n            subscription {\n                messageCount(channel: \"general\", since: \"2024-01-01\")\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(subscription)).toEqual({
            subscription: {
                messageCount: {
                    __args: {
                        channel: "general",
                        since: "2024-01-01",
                    },
                },
            },
        });
    });
    it("Subscription with nested objects and complex arguments", function () {
        var subscription = "\n            subscription NotificationSubscription($input: NotificationInput!) {\n                notificationAdded(input: $input) {\n                    id\n                    message\n                    user {\n                        id\n                        name\n                        profile {\n                            avatar\n                        }\n                    }\n                    metadata {\n                        priority\n                        tags\n                    }\n                }\n            }\n        ";
        var result = (0, index_1.graphQlQueryToJson)(subscription, {
            variables: {
                input: {
                    userId: "123",
                    channels: ["email", "push"],
                    filters: {
                        priority: "high",
                        categories: ["system", "user"],
                    },
                },
            },
        });
        expect(result).toEqual({
            subscription: {
                notificationAdded: {
                    __args: {
                        input: {
                            userId: "123",
                            channels: ["email", "push"],
                            filters: {
                                priority: "high",
                                categories: ["system", "user"],
                            },
                        },
                    },
                    id: true,
                    message: true,
                    user: {
                        id: true,
                        name: true,
                        profile: {
                            avatar: true,
                        },
                    },
                    metadata: {
                        priority: true,
                        tags: true,
                    },
                },
            },
        });
    });
    it("Subscription with directives (directives are ignored)", function () {
        var subscription = "\n            subscription MessageSubscription($includeUser: Boolean!) {\n                messageAdded {\n                    id\n                    content\n                    user @include(if: $includeUser) {\n                        name\n                    }\n                }\n            }\n        ";
        var result = (0, index_1.graphQlQueryToJson)(subscription, {
            variables: {
                includeUser: true,
            },
        });
        expect(result).toEqual({
            subscription: {
                messageAdded: {
                    id: true,
                    content: true,
                    user: {
                        name: true,
                    },
                },
            },
        });
    });
    it("Should throw error for subscription with named fragments", function () {
        var subscriptionWithFragment = "\n            subscription {\n                messageAdded {\n                    ...MessageFields\n                }\n            }\n\n            fragment MessageFields on Message {\n                id\n                content\n                user {\n                    name\n                }\n            }\n        ";
        expect(function () {
            (0, index_1.graphQlQueryToJson)(subscriptionWithFragment);
        }).toThrow("The parsed query has more than one set of definitions");
    });
    it("Should handle subscription with inline fragments", function () {
        var subscriptionWithInlineFragment = "\n            subscription {\n                messageAdded {\n                    id\n                    content\n                    ... on TextMessage {\n                        text\n                    }\n                    ... on ImageMessage {\n                        imageUrl\n                        caption\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(subscriptionWithInlineFragment)).toEqual({
            subscription: {
                messageAdded: {
                    id: true,
                    content: true,
                    __on: [
                        {
                            __typeName: "TextMessage",
                            text: true,
                        },
                        {
                            __typeName: "ImageMessage",
                            imageUrl: true,
                            caption: true,
                        },
                    ],
                },
            },
        });
    });
});
describe("Aliases", function () {
    it("Simple example with aliases", function () {
        var query = "\n            query {\n                viewer {\n                    thingOne {\n                        name\n                        team\n                    }\n                    renamed: thingOne {\n                        propertyC\n                    }\n                }\n            }\n        ";
        var json = (0, index_1.graphQlQueryToJson)(query);
        expect(json).toEqual({
            query: {
                viewer: {
                    thingOne: {
                        name: true,
                        team: true,
                    },
                    renamed: {
                        __aliasFor: "thingOne",
                        propertyC: true,
                    },
                },
            },
        });
    });
});
describe("Enum Types", function () {
    it("Simple enum type", function () {
        var query = "\n            query {\n                viewer {\n                    stuffWithArguments(argumentOne: ALL) {\n                        personalEnumData\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    stuffWithArguments: {
                        __args: {
                            argumentOne: new json_to_graphql_query_1.EnumType("ALL"),
                        },
                        personalEnumData: true,
                    },
                },
            },
        });
    });
    it("Argument is a variable", function () {
        var query = "\n    mutation {\n        getUsers(count: $count) {\n            personal {\n                count\n            }\n        }\n    }";
        var result = (0, index_1.graphQlQueryToJson)(query, {
            variables: {
                count: 1000,
            },
        });
        expect(result).toEqual({
            mutation: {
                getUsers: {
                    __args: {
                        count: 1000,
                    },
                    personal: {
                        count: true,
                    },
                },
            },
        });
    });
});
describe("Complex examples", function () {
    it("Query using name, variables, enums and aliases", function () {
        var variables = { ownership: "ALL", name: "" };
        var query = "query GetAuthenticationsPrivate($name: String, $ownership: OwnershipQueryType!) {\n  viewer {\n    userAuthentications(criteria: {name: $name, ownershipQueryType: $ownership}) {\n      edges {\n        node {\n          id\n          name\n          created\n          creator {\n            name\n            id\n            __typename\n          }\n          service {\n            icon\n            __typename\n          }\n          owner {\n            ownerType\n            __typename\n          }\n          customFields\n          scopes\n          __typename\n        }\n        __typename\n      }\n      __typename\n    }\n    count: userAuthentications(criteria: {ownershipQueryType: ALL}) {\n      edges {\n        node {\n          id\n          __typename\n        }\n        __typename\n      }\n      __typename\n    }\n    __typename\n  }\n}\n";
        var result = (0, index_1.graphQlQueryToJson)(query, { variables: variables });
        var userAuthenticationsQuery = {
            __args: {
                criteria: {
                    name: "",
                    ownershipQueryType: "ALL",
                },
            },
            edges: {
                node: {
                    id: true,
                    name: true,
                    created: true,
                    creator: {
                        name: true,
                        id: true,
                        __typename: true,
                    },
                    service: {
                        icon: true,
                        __typename: true,
                    },
                    owner: {
                        ownerType: true,
                        __typename: true,
                    },
                    customFields: true,
                    scopes: true,
                    __typename: true,
                },
                __typename: true,
            },
            __typename: true,
        };
        var countQuery = {
            __aliasFor: "userAuthentications",
            __args: {
                criteria: {
                    ownershipQueryType: "ALL",
                },
            },
            edges: {
                node: {
                    id: true,
                    __typename: true,
                },
                __typename: true,
            },
            __typename: true,
        };
        var expectedQueryOutput = {
            query: {
                viewer: {
                    __typename: true,
                    userAuthentications: userAuthenticationsQuery,
                    count: countQuery,
                },
            },
        };
        expect(result).toEqual(expectedQueryOutput);
    });
});
describe("Errors", function () {
    it("Throws error if given invalid graphQL schema", function () {
        expect(function () { return (0, index_1.graphQlQueryToJson)("query { asdf sd"); }).toThrow();
    });
    it("Throws error if query has variable which is not passed in", function () {
        var query = "\n        query GetThisStuff($name: String, $lastName: String) {\n            viewer {\n                personal(criteria: {\n                    name: $name,\n                    lastName: $lastName\n                }) {\n                    name\n                    address\n                }\n            }\n        }\n        ";
        var getResult = function () {
            return (0, index_1.graphQlQueryToJson)(query, {
                variables: {
                    name: "PETER",
                },
            });
        };
        expect(getResult).toThrow();
    });
    it("Should throw error when query has multiple operations", function () {
        var multipleOperationsQuery = "\nquery GetUser {\n    user {\n        name\n    }\n}\n\nquery GetPosts {\n    posts {\n        title\n    }\n}\n";
        expect(function () {
            (0, index_1.graphQlQueryToJson)(multipleOperationsQuery);
        }).toThrow("The parsed query has more than one set of definitions");
    });
    it("Should throw error when mixing queries and mutations", function () {
        var mixedOperationsQuery = "\nquery GetUser {\n    user {\n        name\n    }\n}\n\nmutation CreatePost {\n    createPost(title: \"Test\") {\n        id\n    }\n}\n";
        expect(function () {
            (0, index_1.graphQlQueryToJson)(mixedOperationsQuery);
        }).toThrow("The parsed query has more than one set of definitions");
    });
});
describe("Edge Cases and Additional Coverage", function () {
    it("Handles query with no variable definitions", function () {
        var query = "\n            query {\n                viewer {\n                    name\n                }\n            }\n        ";
        expect(function () { return (0, index_1.graphQlQueryToJson)(query, { variables: {} }); }).not.toThrow();
    });
    it("Handles boolean arguments", function () {
        var query = "\n            query {\n                viewer {\n                    posts(published: true, featured: false) {\n                        title\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    posts: {
                        __args: {
                            published: true,
                            featured: false,
                        },
                        title: true,
                    },
                },
            },
        });
    });
    it("Handles float arguments as numbers", function () {
        var query = "\n            query {\n                viewer {\n                    products(rating: 4.5, price: 99.99) {\n                        name\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    products: {
                        __args: {
                            rating: 4.5,
                            price: 99.99,
                        },
                        name: true,
                    },
                },
            },
        });
    });
    it("Handles mixed int and float arguments", function () {
        var query = "\n            query {\n                viewer {\n                    analytics(count: 100, threshold: 2.5, limit: 50, percentage: 87.3) {\n                        data\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    analytics: {
                        __args: {
                            count: 100,
                            threshold: 2.5,
                            limit: 50,
                            percentage: 87.3,
                        },
                        data: true,
                    },
                },
            },
        });
    });
    it("Handles negative float arguments", function () {
        var query = "\n            query {\n                viewer {\n                    adjustments(offset: -12.5, delta: -0.001) {\n                        result\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    adjustments: {
                        __args: {
                            offset: -12.5,
                            delta: -0.001,
                        },
                        result: true,
                    },
                },
            },
        });
    });
    it("Handles zero and decimal edge cases", function () {
        var query = "\n            query {\n                viewer {\n                    measurements(zero: 0.0, decimal: 0.5, trailing: 3.0) {\n                        values\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    measurements: {
                        __args: {
                            zero: 0.0,
                            decimal: 0.5,
                            trailing: 3.0,
                        },
                        values: true,
                    },
                },
            },
        });
    });
    it("Handles floats in nested object arguments", function () {
        var query = "\n            query {\n                viewer {\n                    complexSearch(criteria: {\n                        score: 8.5,\n                        weight: 1.2,\n                        coordinates: {\n                            lat: 40.7128,\n                            lng: -74.0060\n                        }\n                    }) {\n                        results\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    complexSearch: {
                        __args: {
                            criteria: {
                                score: 8.5,
                                weight: 1.2,
                                coordinates: {
                                    lat: 40.7128,
                                    lng: -74.006,
                                },
                            },
                        },
                        results: true,
                    },
                },
            },
        });
    });
    it("Handles floats in array arguments", function () {
        var query = "\n            query {\n                viewer {\n                    dataPoints(values: [1.1, 2.2, 3.3, 4.4]) {\n                        summary\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    dataPoints: {
                        __args: {
                            values: ["1.1", "2.2", "3.3", "4.4"],
                        },
                        summary: true,
                    },
                },
            },
        });
    });
    it("Handles mixed type arrays with floats", function () {
        var query = "\n            query {\n                viewer {\n                    mixedData(values: [1, 2.5, 3, 4.75]) {\n                        result\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    mixedData: {
                        __args: {
                            values: ["1", "2.5", "3", "4.75"],
                        },
                        result: true,
                    },
                },
            },
        });
    });
    it("Handles float variables in arguments", function () {
        var query = "\n            query TestFloatVars($price: Float!, $discount: Float) {\n                viewer {\n                    products(maxPrice: $price, discountRate: $discount) {\n                        name\n                        finalPrice\n                    }\n                }\n            }\n        ";
        var result = (0, index_1.graphQlQueryToJson)(query, {
            variables: {
                price: 199.99,
                discount: 0.15,
            },
        });
        expect(result).toEqual({
            query: {
                viewer: {
                    products: {
                        __args: {
                            maxPrice: 199.99,
                            discountRate: 0.15,
                        },
                        name: true,
                        finalPrice: true,
                    },
                },
            },
        });
    });
    it("Handles scientific notation floats", function () {
        var query = "\n            query {\n                viewer {\n                    scientificData(small: 1e-5, large: 2.5e3, negative: -1.5e-2) {\n                        results\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    scientificData: {
                        __args: {
                            small: 1e-5,
                            large: 2.5e3,
                            negative: -1.5e-2,
                        },
                        results: true,
                    },
                },
            },
        });
    });
    it("Handles null arguments as undefined", function () {
        var query = "\n            query {\n                viewer {\n                    profile(avatar: null) {\n                        name\n                    }\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                viewer: {
                    profile: {
                        __args: {
                            avatar: undefined,
                        },
                        name: true,
                    },
                },
            },
        });
    });
    it("Handles simple string arrays in arguments", function () {
        var query = "\n            query {\n                search(tags: [\"tech\", \"news\"]) {\n                    results\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                search: {
                    __args: {
                        tags: ["tech", "news"],
                    },
                    results: true,
                },
            },
        });
    });
    it("Handles variables with mixed types", function () {
        var query = "\n            query TestQuery($text: String!, $count: Int!, $active: Boolean!, $rating: Float!) {\n                search(filter: {\n                    text: $text,\n                    count: $count,\n                    active: $active,\n                    rating: $rating\n                }) {\n                    results\n                }\n            }\n        ";
        var result = (0, index_1.graphQlQueryToJson)(query, {
            variables: {
                text: "test",
                count: 5,
                active: true,
                rating: 3.7,
            },
        });
        expect(result).toEqual({
            query: {
                search: {
                    __args: {
                        filter: {
                            text: "test",
                            count: 5,
                            active: true,
                            rating: 3.7,
                        },
                    },
                    results: true,
                },
            },
        });
    });
    it("Handles simple variable replacement in nested objects", function () {
        var query = "\n            query TestQuery($name: String!) {\n                user(filter: { name: $name, active: true }) {\n                    id\n                    name\n                }\n            }\n        ";
        var result = (0, index_1.graphQlQueryToJson)(query, {
            variables: {
                name: "Alice",
            },
        });
        expect(result).toEqual({
            query: {
                user: {
                    __args: {
                        filter: {
                            name: "Alice",
                            active: true,
                        },
                    },
                    id: true,
                    name: true,
                },
            },
        });
    });
    it("Handles enums in lists", function () {
        var query = "\n            query {\n                posts(statuses: [PUBLISHED, DRAFT, ARCHIVED]) {\n                    title\n                }\n            }\n        ";
        expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
            query: {
                posts: {
                    __args: {
                        statuses: ["PUBLISHED", "DRAFT", "ARCHIVED"],
                    },
                    title: true,
                },
            },
        });
    });
    it("Handles string concatenation edge case in variable replacement", function () {
        var query = "\n            query TestQuery($prefix: String!) {\n                search(term: $prefix) {\n                    results\n                }\n            }\n        ";
        var result = (0, index_1.graphQlQueryToJson)(query, {
            variables: {
                prefix: "test_prefix_value",
            },
        });
        expect(result).toEqual({
            query: {
                search: {
                    __args: {
                        term: "test_prefix_value",
                    },
                    results: true,
                },
            },
        });
    });
    describe("Inline Fragments", function () {
        it("Single inline fragment", function () {
            var query = "\n                query {\n                    posts {\n                        title\n                        ... on TextPost {\n                            content\n                        }\n                    }\n                }\n            ";
            expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
                query: {
                    posts: {
                        title: true,
                        __on: {
                            __typeName: "TextPost",
                            content: true,
                        },
                    },
                },
            });
        });
        it("Multiple inline fragments", function () {
            var query = "\n                query {\n                    posts {\n                        title\n                        ... on TextPost {\n                            content\n                            wordCount\n                        }\n                        ... on ImagePost {\n                            imageUrl\n                            altText\n                        }\n                    }\n                }\n            ";
            expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
                query: {
                    posts: {
                        title: true,
                        __on: [
                            {
                                __typeName: "TextPost",
                                content: true,
                                wordCount: true,
                            },
                            {
                                __typeName: "ImagePost",
                                imageUrl: true,
                                altText: true,
                            },
                        ],
                    },
                },
            });
        });
        it("Inline fragment with nested selections", function () {
            var query = "\n                query {\n                    posts {\n                        title\n                        ... on TextPost {\n                            content\n                            author {\n                                name\n                                bio\n                            }\n                        }\n                    }\n                }\n            ";
            expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
                query: {
                    posts: {
                        title: true,
                        __on: {
                            __typeName: "TextPost",
                            content: true,
                            author: {
                                name: true,
                                bio: true,
                            },
                        },
                    },
                },
            });
        });
        it("Inline fragment with arguments", function () {
            var query = "\n                query {\n                    posts {\n                        title\n                        ... on TextPost {\n                            content\n                            comments(limit: 5) {\n                                text\n                            }\n                        }\n                    }\n                }\n            ";
            expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
                query: {
                    posts: {
                        title: true,
                        __on: {
                            __typeName: "TextPost",
                            content: true,
                            comments: {
                                __args: {
                                    limit: 5,
                                },
                                text: true,
                            },
                        },
                    },
                },
            });
        });
        it("Nested inline fragments", function () {
            var query = "\n                query {\n                    media {\n                        ... on Post {\n                            title\n                            ... on TextPost {\n                                content\n                            }\n                            ... on ImagePost {\n                                imageUrl\n                            }\n                        }\n                        ... on Comment {\n                            text\n                            author {\n                                name\n                            }\n                        }\n                    }\n                }\n            ";
            expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
                query: {
                    media: {
                        __on: [
                            {
                                __typeName: "Post",
                                title: true,
                                __on: [
                                    {
                                        __typeName: "TextPost",
                                        content: true,
                                    },
                                    {
                                        __typeName: "ImagePost",
                                        imageUrl: true,
                                    },
                                ],
                            },
                            {
                                __typeName: "Comment",
                                text: true,
                                author: {
                                    name: true,
                                },
                            },
                        ],
                    },
                },
            });
        });
        it("Inline fragment with enum arguments", function () {
            var query = "\n                query {\n                    posts {\n                        title\n                        ... on TextPost {\n                            content(format: MARKDOWN) {\n                                rendered\n                            }\n                        }\n                    }\n                }\n            ";
            expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
                query: {
                    posts: {
                        title: true,
                        __on: {
                            __typeName: "TextPost",
                            content: {
                                __args: {
                                    format: new json_to_graphql_query_1.EnumType("MARKDOWN"),
                                },
                                rendered: true,
                            },
                        },
                    },
                },
            });
        });
        it("Inline fragment with variables", function () {
            var query = "\n                query GetPosts($limit: Int!) {\n                    posts {\n                        title\n                        ... on TextPost {\n                            comments(limit: $limit) {\n                                text\n                            }\n                        }\n                    }\n                }\n            ";
            expect((0, index_1.graphQlQueryToJson)(query, { variables: { limit: 10 } })).toEqual({
                query: {
                    posts: {
                        title: true,
                        __on: {
                            __typeName: "TextPost",
                            comments: {
                                __args: {
                                    limit: 10,
                                },
                                text: true,
                            },
                        },
                    },
                },
            });
        });
        it("Inline fragment with aliases", function () {
            var query = "\n                query {\n                    posts {\n                        title\n                        ... on TextPost {\n                            textContent: content\n                            authorName: author {\n                                name\n                            }\n                        }\n                    }\n                }\n            ";
            expect((0, index_1.graphQlQueryToJson)(query)).toEqual({
                query: {
                    posts: {
                        title: true,
                        __on: {
                            __typeName: "TextPost",
                            textContent: {
                                __aliasFor: "content",
                            },
                            authorName: {
                                __aliasFor: "author",
                                name: true,
                            },
                        },
                    },
                },
            });
        });
    });
});
