// graphql-query-to-json/dist/test/readme_examples.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("../index");
describe("README Examples Validation", function () {
    describe("Simple Queries", function () {
        it("Single and multiple fields", function () {
            var query = "\nquery {\n    viewer {\n        name\n        email\n    }\n    user {\n        profile {\n            avatar\n            bio\n        }\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(query);
            expect(result).toEqual({
                query: {
                    viewer: {
                        name: true,
                        email: true,
                    },
                    user: {
                        profile: {
                            avatar: true,
                            bio: true,
                        },
                    },
                },
            });
        });
    });
    describe("Variables and Arguments", function () {
        it("Query with variables", function () {
            var query = "\nquery GetUser($userId: ID!, $includeProfile: Boolean!) {\n    user(id: $userId) {\n        name\n        email\n        profile @include(if: $includeProfile) {\n            bio\n            avatar\n        }\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(query, {
                variables: {
                    userId: "123",
                    includeProfile: true,
                },
            });
            expect(result).toEqual({
                query: {
                    user: {
                        __args: { id: "123" },
                        name: true,
                        email: true,
                        profile: {
                            bio: true,
                            avatar: true,
                        },
                    },
                },
            });
        });
    });
    describe("Scalar Fields with Arguments", function () {
        it("Scalar fields that accept arguments", function () {
            var query = "\nquery {\n    userCount(filter: \"active\")\n    totalRevenue(currency: \"USD\", year: 2024)\n    averageRating(precision: 2)\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(query);
            expect(result).toEqual({
                query: {
                    userCount: {
                        __args: { filter: "active" },
                    },
                    totalRevenue: {
                        __args: { currency: "USD", year: 2024 },
                    },
                    averageRating: {
                        __args: { precision: 2 },
                    },
                },
            });
        });
    });
    describe("Mutations with Simple Arguments", function () {
        it("Simple mutation using string argument with sibling queries", function () {
            var mutation = "\n        mutation {\n            getPersonalStuff(name: \"PETER\") {\n                personal {\n                    name\n                    address\n                }\n                other {\n                    parents\n                }\n            }\n        }\n        ";
            expect((0, index_1.graphQlQueryToJson)(mutation)).toEqual({
                mutation: {
                    getPersonalStuff: {
                        __args: {
                            name: "PETER",
                        },
                        personal: {
                            name: true,
                            address: true,
                        },
                        other: {
                            parents: true,
                        },
                    },
                },
            });
        });
    });
    describe("Mutations with Complex Arguments", function () {
        it("Mutation with nested object arguments", function () {
            var mutation = "\nmutation CreateUser($input: UserInput!) {\n    createUser(input: $input) {\n        id\n        name\n        profile {\n            email\n            settings {\n                theme\n                notifications\n            }\n        }\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(mutation, {
                variables: {
                    input: {
                        name: "John Doe",
                        email: "john@example.com",
                    },
                },
            });
            expect(result).toEqual({
                mutation: {
                    createUser: {
                        __args: {
                            input: {
                                name: "John Doe",
                                email: "john@example.com",
                            },
                        },
                        id: true,
                        name: true,
                        profile: {
                            email: true,
                            settings: {
                                theme: true,
                                notifications: true,
                            },
                        },
                    },
                },
            });
        });
    });
    describe("Aliases and Field Renaming", function () {
        it("Multiple aliases for the same field", function () {
            var query = "\nquery {\n    currentUser: user(id: 1) {\n        name\n        email\n    }\n    adminUser: user(id: 2) {\n        name\n        permissions\n    }\n    guestUser: user(id: 3) {\n        name\n        status\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(query);
            expect(result).toEqual({
                query: {
                    currentUser: {
                        __aliasFor: "user",
                        __args: { id: 1 },
                        name: true,
                        email: true,
                    },
                    adminUser: {
                        __aliasFor: "user",
                        __args: { id: 2 },
                        name: true,
                        permissions: true,
                    },
                    guestUser: {
                        __aliasFor: "user",
                        __args: { id: 3 },
                        name: true,
                        status: true,
                    },
                },
            });
        });
    });
    describe("Enum Types", function () {
        it("Enums in arguments", function () {
            var query = "\nquery {\n    posts(status: PUBLISHED, orderBy: CREATED_DESC) {\n        title\n        content\n    }\n    users(role: ADMIN, sortBy: NAME_ASC) {\n        name\n        email\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(query);
            expect(result).toEqual({
                query: {
                    posts: {
                        __args: {
                            status: { value: "PUBLISHED" },
                            orderBy: { value: "CREATED_DESC" },
                        },
                        title: true,
                        content: true,
                    },
                    users: {
                        __args: {
                            role: { value: "ADMIN" },
                            sortBy: { value: "NAME_ASC" },
                        },
                        name: true,
                        email: true,
                    },
                },
            });
        });
    });
    describe("Array Arguments", function () {
        it("Lists and arrays as arguments", function () {
            var mutation = "\nmutation {\n    updateUser(\n        id: \"123\",\n        tags: [\"developer\", \"typescript\", \"graphql\"],\n        permissions: [READ, WRITE, ADMIN]\n    ) {\n        id\n        tags\n        permissions\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(mutation);
            expect(result).toEqual({
                mutation: {
                    updateUser: {
                        __args: {
                            id: "123",
                            tags: ["developer", "typescript", "graphql"],
                            permissions: ["READ", "WRITE", "ADMIN"],
                        },
                        id: true,
                        tags: true,
                        permissions: true,
                    },
                },
            });
        });
    });
    describe("Empty Values and Edge Cases", function () {
        it("Empty strings, objects, and arrays", function () {
            var mutation = "\nmutation {\n    createRecord(input: {\n        name: \"\",\n        metadata: {},\n        tags: [],\n        count: 0,\n        isActive: false\n    }) {\n        id\n        status\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(mutation);
            expect(result).toEqual({
                mutation: {
                    createRecord: {
                        __args: {
                            input: {
                                name: "",
                                metadata: {},
                                tags: [],
                                count: 0,
                                isActive: false,
                            },
                        },
                        id: true,
                        status: true,
                    },
                },
            });
        });
    });
    describe("Deeply Nested Objects", function () {
        it("Complex nested structures", function () {
            var query = "\nquery {\n    organization {\n        teams {\n            members {\n                user {\n                    profile {\n                        personalInfo {\n                            address {\n                                street\n                                city\n                                country\n                            }\n                        }\n                    }\n                }\n            }\n        }\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(query);
            expect(result).toEqual({
                query: {
                    organization: {
                        teams: {
                            members: {
                                user: {
                                    profile: {
                                        personalInfo: {
                                            address: {
                                                street: true,
                                                city: true,
                                                country: true,
                                            },
                                        },
                                    },
                                },
                            },
                        },
                    },
                },
            });
        });
    });
    describe("Mixed Variable Types", function () {
        it("Various data types as variables including floats", function () {
            var query = "\nquery SearchContent(\n    $text: String!,\n    $limit: Int!,\n    $rating: Float!,\n    $price: Float,\n    $filters: FilterInput!,\n    $includeArchived: Boolean\n) {\n    search(\n        query: $text,\n        first: $limit,\n        minRating: $rating,\n        maxPrice: $price,\n        filters: $filters,\n        archived: $includeArchived\n    ) {\n        results {\n            id\n            title\n            excerpt\n        }\n        totalCount\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(query, {
                variables: {
                    text: "GraphQL tutorial",
                    limit: 10,
                    rating: 4.8,
                    price: 29.99,
                    filters: {
                        category: "tutorial",
                        difficulty: "beginner",
                        tags: ["graphql", "api"],
                    },
                    includeArchived: false,
                },
            });
            expect(result).toEqual({
                query: {
                    search: {
                        __args: {
                            query: "GraphQL tutorial",
                            first: 10,
                            minRating: 4.8,
                            maxPrice: 29.99,
                            filters: {
                                category: "tutorial",
                                difficulty: "beginner",
                                tags: ["graphql", "api"],
                            },
                            archived: false,
                        },
                        results: {
                            id: true,
                            title: true,
                            excerpt: true,
                        },
                        totalCount: true,
                    },
                },
            });
        });
        it("Float arguments and numeric types", function () {
            var query = "\nquery GetProducts {\n    products(\n        minRating: 4.5,\n        maxPrice: 99.99,\n        discount: -10.5,\n        threshold: 0.001,\n        scientific: 2.5e3\n    ) {\n        name\n        rating\n        price\n    }\n    analytics(\n        coordinates: {\n            lat: 40.7128,\n            lng: -74.006\n        },\n        mixed: [1, 2.5, 3, 4.75]\n    ) {\n        data\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(query);
            expect(result).toEqual({
                query: {
                    products: {
                        __args: {
                            minRating: 4.5,
                            maxPrice: 99.99,
                            discount: -10.5,
                            threshold: 0.001,
                            scientific: 2500,
                        },
                        name: true,
                        rating: true,
                        price: true,
                    },
                    analytics: {
                        __args: {
                            coordinates: {
                                lat: 40.7128,
                                lng: -74.006,
                            },
                            mixed: ["1", "2.5", "3", "4.75"],
                        },
                        data: true,
                    },
                },
            });
        });
    });
    describe("Subscriptions", function () {
        it("Basic subscription", function () {
            var subscription = "\nsubscription {\n    messageAdded {\n        id\n        content\n        user {\n            name\n            email\n        }\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(subscription);
            expect(result).toEqual({
                subscription: {
                    messageAdded: {
                        id: true,
                        content: true,
                        user: {
                            name: true,
                            email: true,
                        },
                    },
                },
            });
        });
        it("Subscription with variables and arguments", function () {
            var subscription = "\nsubscription MessageSubscription($userId: ID!, $channel: String!) {\n    messageAdded(userId: $userId, channel: $channel) {\n        id\n        content\n        timestamp\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(subscription, {
                variables: {
                    userId: "123",
                    channel: "general",
                },
            });
            expect(result).toEqual({
                subscription: {
                    messageAdded: {
                        __args: {
                            userId: "123",
                            channel: "general",
                        },
                        id: true,
                        content: true,
                        timestamp: true,
                    },
                },
            });
        });
        it("Subscription with aliases and enums", function () {
            var subscription = "\nsubscription {\n    latestMessage: messageAdded(channel: PUBLIC) {\n        id\n        content\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(subscription);
            expect(result).toEqual({
                subscription: {
                    latestMessage: {
                        __aliasFor: "messageAdded",
                        __args: {
                            channel: { value: "PUBLIC" },
                        },
                        id: true,
                        content: true,
                    },
                },
            });
        });
    });
    describe("Inline Fragments", function () {
        it("Single inline fragment", function () {
            var query = "\nquery {\n    posts {\n        title\n        ... on TextPost {\n            content\n            wordCount\n        }\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(query);
            expect(result).toEqual({
                query: {
                    posts: {
                        title: true,
                        __on: {
                            __typeName: "TextPost",
                            content: true,
                            wordCount: true,
                        },
                    },
                },
            });
        });
        it("Multiple inline fragments", function () {
            var query = "\nquery {\n    media {\n        ... on TextPost {\n            content\n            author {\n                name\n            }\n        }\n        ... on ImagePost {\n            imageUrl\n            altText\n        }\n        ... on VideoPost {\n            videoUrl\n            duration\n        }\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(query);
            expect(result).toEqual({
                query: {
                    media: {
                        __on: [
                            {
                                __typeName: "TextPost",
                                content: true,
                                author: {
                                    name: true,
                                },
                            },
                            {
                                __typeName: "ImagePost",
                                imageUrl: true,
                                altText: true,
                            },
                            {
                                __typeName: "VideoPost",
                                videoUrl: true,
                                duration: true,
                            },
                        ],
                    },
                },
            });
        });
        it("Inline fragments with arguments and variables", function () {
            var query = "\nquery GetPosts($limit: Int!) {\n    posts {\n        title\n        ... on TextPost {\n            comments(limit: $limit) {\n                text\n                author {\n                    name\n                }\n            }\n        }\n    }\n}\n";
            var result = (0, index_1.graphQlQueryToJson)(query, {
                variables: { limit: 5 },
            });
            expect(result).toEqual({
                query: {
                    posts: {
                        title: true,
                        __on: {
                            __typeName: "TextPost",
                            comments: {
                                __args: { limit: 5 },
                                text: true,
                                author: {
                                    name: true,
                                },
                            },
                        },
                    },
                },
            });
        });
    });
});
